### PR TITLE
fix: Implement random reader for generating data.

### LINKED ISSUE #742: Functional tests should use streaming
Minio-go functional tests will be used in Mint framework to test Minio server. This calls for uniform usage of Files across Minio-SDK functional tests.

It was concluded that (https://github.com/minio/mint/pull/67/files)

- Tests should generate their own data by using input/output streams. These streams should be used for testing upload cases.

This makes tests flexible, removes too much dependency on /mint/data directory and reduces the size of `mint` Docker image.

Once these changes are done in minio-go, `mint` will be updated to use these test cases instead of maintaining copy of minio-go functional tests.

### PR DESCRIPTION
**Pull request by harshavardhana**
Fixes #742

### CODE REVIEW

File: functional_tests.go [hunk 1]
```diff
@@ -85,14 +85,46 @@ func getFilePath(filename string) (filepath string) {
...
 
+// Implements a type of reader which returns a
+// constant stream of random content
+type randomReader struct {
+	TotalBytes       int64
+	RandomSeedChar   rune
+	currentReadBytes int
+}
+
+func (r *randomReader) Read(p []byte) (n int, err error) {
```
Comment (vadmeste) [inline functional_tests.go]: Can you add some comments in this function ? @harshavardhana, it is quite complicated to read

File: functional_tests.go [hunk 2]
```diff
@@ -255,20 +287,21 @@ func testPutObjectReadAt() {
...
 
-	n, err := c.PutObject(bucketName, objectName, bytes.NewReader(buf), objectContentType)
+	n, err := c.PutObject(bucketName, objectName, reader, objectContentType)
 	if err != nil {
 		logger().Fatal("Error:", err, bucketName, objectName)
 	}
 
-	if n != int64(len(buf)) {
-		logger().Fatalf("Error: number of bytes does not match, want %v, got %v\n", len(buf), n)
+	if n != int64(MinPartSize) {
```
Comment (vadmeste) [inline functional_tests.go]: If we are not using a random data source, then we get our data from 'datafile-65-MB' which means 65MB and not MinPartSize (64MB)

Comment (harshavardhana) [inline functional_tests.go]: Yes you are right .. looks like it was an existing problem too @vadmeste

File: functional_tests.go (hunk 1 above)
Comment (harshavardhana) [inline functional_tests.go]: Sure.. will do.

File: functional_tests.go [hunk 3]
```diff
@@ -85,14 +87,59 @@ func getFilePath(filename string) (filepath string) {
...
 // read data from file if it exists or optionally create a buffer of particular size
-func getDataBuffer(fileName string, size int) (buf []byte) {
+func getDataReader(fileName string, size int) io.ReadCloser {
 	if _, err := os.Stat(getFilePath(fileName)); os.IsNotExist(err) {
-		buf = bytes.Repeat([]byte(string('a')), size)
-		return
+		return &randomReader{
+			TotalBytes:     int64(size),
+			RandomSeedChar: "a",
+		}
```
Comment (krisis) [inline functional_tests.go]: we could remove code for counting bytes read by composing a simple random reader with `io.LimitReader`.
E.g,
```go
  return io.LimitReader{R: randomReader{seed: "a"}, N: size}
```
where

```go
type randomReader struct {
  seed string
}

func (r *randomReader) Read (b []byte) (int, error) {
   return copy(b, bytes.Repeat(r.seed, len(b))), nil
}
```

Comment (harshavardhana) [inline functional_tests.go]: done

File: functional_tests.go [hunk 4]
```diff
@@ -883,26 +920,24 @@ func testFPutObjectMultipart() {
...
 	}
-	totalSize := MinPartSize * 1
+	totalSize := sixtyFiveMiB * 1
 	// Set base object name
 	objectName := bucketName + "FPutObject"
 	objectContentType := "testapplication/octet-stream"
 
 	// Perform standard FPutObject with contentType provided (Expecting application/octet-stream)
-	n, err := c.FPutObject(bucketName, objectName+"-standard", fileName, objectContentType)
+	ny, err := c.FPutObject(bucketName, objectName+"-standard", fileName, objectContentType)
```
Comment (krisis) [inline functional_tests.go]: looks like a typo. s/ny/n

**Comment by vadmeste** [comment]
@harshavardhana, it looks like functional tests are not passing

```
$ go run functional_tests.go
...
FATA[0032] Error:Put http://localhost:9000/minio-go-testa0mxfe1pv2yymobng/minio-go-testa0mxfe1pv2yymobngFPutObject-standard?partNumber=1&uploadId=ee43d6b8-bc7e-4f3b-b6bb-6a634459d6b9: Connection closed by foreign host http://localhost:9000/minio-go-testa0mxfe1pv2yymobng/minio-go-testa0mxfe1pv2yymobngFPutObject-standard?partNumber=1&uploadId=ee43d6b8-bc7e-4f3b-b6bb-6a634459d6b9. Retry again.  file=functional_tests.go function:=main.testFPutObject line#=1038

...
```

Minio server log:
```
ERRO[0027] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0027] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0029] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0032] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0036] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0567] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0567] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0569] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0572] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm
ERRO[0578] Unable to create object part.                 cause=#has incomplete body source=[object-handlers.go:850:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:547:fsObjects.PutObjectPart <autogenerated>:388:(*fsObjects).PutObjectPart object-handlers.go:828:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm

```

**Comment by harshavardhana** [comment]
Already opened a bug for this @vadmeste #784 it has nothing to do with this PR.

**Comment by harshavardhana** [comment]
@krisis any reviews?

**Pull request merged**